File: WEEK7/Js/bj8_2098.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const w = [];
for (let i = 1; i <= n; i++) {
    w.push(input[i].split(' ').map(Number));
}

const INF = 16000000;
const dp = Array.from({ length: n }, () => new Int32Array(1 << n).fill(-1));

function tsp(current, visited) {
    if (visited === (1 << n) - 1) {
        return w[current][0] === 0 ? INF : w[current][0];
    }

    if (dp[current][visited] !== -1) {
        return dp[current][visited];
    }

    dp[current][visited] = INF;

    for (let i = 0; i < n; i++) {
        if ((visited & (1 << i)) === 0 && w[current][i] !== 0) {
            dp[current][visited] = Math.min(
                dp[current][visited],
                tsp(i, visited | (1 << i)) + w[current][i]
            );
        }
    }
    return dp[current][visited];
}

console.log(tsp(0, 1));

File: WEEK7/Python/bj8_2098.py
import sys

n = int(sys.stdin.readline())
w = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
dp = [[-1] * (1 << n) for _ in range(n)]
INF = 16000000


def tsp(current, visited):
    if visited == (1 << n) - 1:
        return w[current][0] if w[current][0] > 0 else INF

    if dp[current][visited] != -1:
        return dp[current][visited]

    dp[current][visited] = INF
    for i in range(n):
        if not (visited & (1 << i)) and w[current][i] > 0:
            dp[current][visited] = min(
                dp[current][visited], tsp(i, visited | (1 << i)) + w[current][i]
            )

    return dp[current][visited]


print(tsp(0, 1))
